cockatrice/src/client/ui/widgets/deck_analytics/mana_base_widget.cpp

```diff
@@ -0,0 +1,142 @@
+#include "mana_base_widget.h"
+
+#include "../../../../deck/deck_loader.h"
+#include "../../../../game/cards/card_database.h"
+#include "../../../../game/cards/card_database_manager.h"
+#include "../../../../main.h"
+#include "../general/display/banner_widget.h"
+#include "../general/display/bar_widget.h"
+
+#include <QRegularExpression>
+#include <decklist.h>
+#include <regex>
+#include <unordered_map>
+
+ManaBaseWidget::ManaBaseWidget(QWidget *parent, DeckListModel *_deckListModel)
+    : QWidget(parent), deckListModel(_deckListModel)
+{
+    layout = new QVBoxLayout(this);
+    setLayout(layout);
+
+    bannerWidget = new BannerWidget(this, "Mana Base", Qt::Vertical, 100);
+    bannerWidget->setMaximumHeight(100);
+    layout->addWidget(bannerWidget);
+
+    barLayout = new QHBoxLayout();
+    layout->addLayout(barLayout);
+
+    connect(deckListModel, &DeckListModel::dataChanged, this, &ManaBaseWidget::analyzeManaBase);
+}
+
+void ManaBaseWidget::setDeckModel(DeckListModel *deckModel)
+{
+    deckListModel = deckModel;
+    connect(deckListModel, &DeckListModel::dataChanged, this, &ManaBaseWidget::analyzeManaBase);
+    analyzeManaBase();
+}
+
+std::unordered_map<char, int> ManaBaseWidget::analyzeManaBase()
+{
+    manaBaseMap.clear();
+    InnerDecklistNode *listRoot = deckListModel->getDeckList()->getRoot();
+    for (int i = 0; i < listRoot->size(); i++) {
+        InnerDecklistNode *currentZone = dynamic_cast<InnerDecklistNode *>(listRoot->at(i));
+        for (int j = 0; j < currentZone->size(); j++) {
+            DecklistCardNode *currentCard = dynamic_cast<DecklistCardNode *>(currentZone->at(j));
+            if (!currentCard)
+                continue;
+
+            for (int k = 0; k < currentCard->getNumber(); ++k) {
+                CardInfoPtr info = CardDatabaseManager::getInstance()->getCard(currentCard->getName());
+                if (info) {
+                    auto devotion = determineManaProduction(info->getText());
+                    mergeManaCounts(manaBaseMap, devotion);
+                }
+            }
+        }
+    }
+
+    updateDisplay();
+    return manaBaseMap;
+}
+
+void ManaBaseWidget::updateDisplay()
+{
+    // Clear the layout first
+    QLayoutItem *item;
+    while ((item = barLayout->takeAt(0)) != nullptr) {
+        delete item->widget();
```

Comment:
I recall we ran into an issue with doing this. I think we need to item->deleteLater after deleting the inner widget